File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

while True:
    string = sys.stdin.readline().rstrip()
    stack = []
    if string == '.':
        break
    for s in string:
        if s == '(' or s == '[':
            stack.append(s)
        elif s == ')':
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                stack.append('(')
                break
        elif s == ']':
            if stack and stack[-1] == '[':
                stack.pop()
            else:
                stack.append('[')
                break
    print('yes' if not stack else 'no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
	cin.tie(0)->sync_with_stdio(0);

	string s;
	while (true) {
		getline(cin, s);
		if (s == ".") break;
		stack<char> st;
		for (char& c : s) {
			if (c == '(' || c == '[') {
				st.push(c);
			}
			else if (c == ')') {
				if (!st.empty() && st.top() == '(') {
					st.pop();
				}
				else {
					st.push(c);
					break;
				}
			}
			else if (c == ']') {
				if (!st.empty() && st.top() == '[') {
					st.pop();
				}
				else {
					st.push(c);
					break;
				}
			}
		}
		if (st.empty()) cout << "yes" << '\n';
		else cout << "no" << '\n';
	}
}